People management v1 wrap-up

oguzkocer opened the pull request:
Relevant #3901. I've started this PR with the intention of adding the change role & remove user functionality's UI side, but I kept finding small issues that I didn't feel deserved a separate PR. However, it got a bit out of hand and became a bigger PR than I'd have liked. Having said that, this should be the last PR, aside from adding Analytics, and we can wrap up the v1 of this feature once this is merged.

A list of things that was accomplished in this PR:
* Changing role: UI side
* Removing a user: network call, db changes, UI etc.
* Error handling for network calls: showing a toast message when a call fails
* Using custom toolbar to match the other pages' design
* When a new list of users is received from a network call, we delete the previous records - This is done to make sure that if the list is changed on remote we reflect it in the DB.
* Remove the users from people table when a blog is removed from the app
* Many other small improvements overall.

I've already added a bunch of comments to explain what certain parts of the code do, but if there is anything not immediately clear, please let me know so I can add more comments.

To test: Since this is the last PR, we should test the whole flow

Needs review: @hypest

hypest commented on WordPress/src/main/java/org/wordpress/android/datasets/PeopleTable.java:
```diff
@@ -77,10 +80,21 @@ public static void savePeople(List<Person> peopleList) {
         String[] args = { Integer.toString(localTableBlogId) };
-        Cursor c = getReadableDb().rawQuery("SELECT * FROM " + PEOPLE_TABLE + " WHERE local_blog_id=?", args);
+        Cursor c = getReadableDb().rawQuery("SELECT * FROM " + PEOPLE_TABLE +
+                " WHERE local_blog_id=? ORDER BY display_name, user_name", args);
```
Not in this changeset but, there are a couple of lines below this one that could use the same treatment and be wrapped at 120chars. Cheers!

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/people/PersonDetailFragment.java:
```diff
@@ -106,4 +157,104 @@ public void setPersonDetails(long personID, int localTableBlogID) {
     }
+
+    // Checks current user's capabilities to decide whether she can change the role or not
+    @SuppressWarnings("deprecation")
```
This comment is more like a personal preference so feel free to move passed it: I'd recommend factoring out the code that the `SuppressWarnings` is about (I assume it's the `setBackgroundDrawable()` call) into its own function and have the suppression apply to that one.
The idea is to avoid having suppressions on "biggish" functions which might hide potential bugs. I would suggest a `clearRoleContainerBackground()` function or something in this case. Cheers!

hypest commented on WordPress/src/main/res/values/strings.xml:
```diff
@@ -1023,6 +1023,9 @@
     <string name="error_open_list_from_notification">This post or page was published on another site</string>
+    <string name="error_fetch_people_list">Couldn\'t retrieve users for the site</string>
+    <string name="error_update_role">Couldn\'t update role of the user</string>
+    <string name="error_remove_user">Couldn\'t remove user</string>
```
I can certainly be wrong here but, perhaps we can simplify these three strings into:
`Could not retrieve site users`,
`Could not update user role`,
`Could not remove user`.

oguzkocer replied:
I've added these strings temporarily and as you can see from the checklist in the master issue: https://github.com/wordpress-mobile/WordPress-Android/issues/3901#issuecomment-217828361 it's on my list to get feedback on them from designers on the main PR that'll be merged to `develop`. Thanks!

hypest commented on WordPress/src/main/res/layout/people_list_row.xml:
```diff
@@ -1,56 +1,32 @@
+        android:layout_alignParentRight="true"
+        android:layout_marginEnd="@dimen/margin_extra_large"
+        android:layout_marginRight="@dimen/margin_extra_large"
+        android:layout_marginTop="@dimen/margin_extra_large"
```
It seems that the role text could use a bit more top margin to reach the baseline of the person username.

oguzkocer replied:
@hypest You're right that there should be a bit more top margin here. However, as far as I can tell from @rickybanister's design, it shouldn't reach the baseline of the username. Adding 4dp more to the top margin seems to perfectly match the design here: https://cloudup.com/cLPBpxkBBjQ. Let me know if this works for you.

hypest commented on WordPress/src/main/res/layout/person_detail_fragment.xml:
```diff
@@ -1,97 +1,63 @@
-            android:layout_height="@dimen/avatar_sz_large"
-            android:layout_marginLeft="@dimen/margin_extra_large"
-            android:layout_marginTop="@dimen/margin_extra_large" />
+    <LinearLayout
```
It seems that the role elements have exactly the same styling as the person's name ones but, in practice, the role row is tappable. I was a bit confused when I first reached this screen as I didn't have a a clue that I could change the role here. After I found out, I tried to tap the username row to see what will happen (which ofc didn't do something).

So, there seems to be a minor design issue here. I don't consider it a blocker for the PR but perhaps we can improve it. Cheers!

oguzkocer replied:
I don't think there is an issue here since it should be pretty obvious that you can't change the username of a user. I think we're also aiming for consistency with Calypso for the design, so when people come from Calypso, it'll immediately feel familiar. It's still worth pinging @rickybanister since he worked on the design.

hypest replied:
FWIW, in Calypso the Role changing is done through a dropdown and that makes it quite clear that it is modifiable versus the username (which indeed, is not editable in Calypso either). What I'm trying to highlight though is that those two look exactly the same on the PR, not providing enough clue to the fact that one is modifiable while the other is not. @rickybanister , is there some insight I'm missing here? Cheers!

rickybanister replied:
Can I see a visual representation of your concern?

The material guidelines suggest using the list component to open a dialog with select menu options instead of the older 'dropdown' menu pattern. In the list view we are showing the smaller role 'badge'. 

In Calypso, if you don't have permission you won't see that dropdown either.

hypest replied:
> Can I see a visual representation of your concern?

Hey @rickybanister , sure. Have a look at this screenshot of the people details screen in the current PR (annotations by me in red):
![people-mgmt-details-styling-issue](https://cloud.githubusercontent.com/assets/1032913/15268741/dea0ae02-19f1-11e6-8710-56a02ebaec1b.png)

Notice the two boxed areas. One is tappable the other is not but their styling is exactly the same (same background, text colour, text styling).

Not suggesting using a dropdown here, I've only mentioned it to point out that Calypso does differentiate between the person's username title (non tappable) and the user's role widget. In the current Android design of this feature, we mix read-only elements with read-write ones but style them the same. Perhaps we can tweak this a bit to provide more clue and make them more "separation". Lemme know if it still is not clear what i'm referring to. Cheers!

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/people/PersonDetailFragment.java:
```diff
@@ -2,22 +2,36 @@
 import org.wordpress.android.widgets.WPNetworkImageView;
 
-public class PersonDetailFragment extends Fragment {
+public class PersonDetailFragment extends Fragment implements View.OnClickListener {
```
When first read this I though "Hmm, interesting, the fragment acts as a click handler? What does that respond to?!". I then understood that that's wired up to the Role row to trigger the role change dialog.

Can you share more info on this choice? If not a particular one, it's will probably be more clear to provide an dedicated `OnClickListener` to `setOnClickListener()` on the spot. Cheers!

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/people/PersonDetailFragment.java:
```diff
@@ -106,4 +157,104 @@ public void setPersonDetails(long personID, int localTableBlogID) {
+        }
+
+        private void changeSelection(int position) {
+            mSelectedRole = getItem(position);
```
This doesn't seem to work well. The variable is kept between invocations of the Role dialog leading to the previously selected role to be checked, even if the previous dialog was dismissed without "OK".

While at it, does the `mSelectedRole` variable really need to be Fragment global? It seems to belong more into the `RoleListAdapter` and not even be directly exposed as all.

hypest commented:
A difference to similarly looking dialogs in the app is that the role change popup dialog is dismissed on rotate.

hypest commented on WordPress/src/main/res/drawable/people_list_divider.xml:
```diff
@@ -0,0 +1,8 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<inset xmlns:android="http://schemas.android.com/apk/res/android"
+       android:insetLeft="@dimen/people_list_divider_left_margin">
```
Hmm, this seems to adhere to the Material Design specs but we don't put such a margin in the rest of the lists in the app (where we use full width separators). This difference make this one look kinda strange to my eyes.

Can we perhaps defer this change to a future release, where we will change all suitable separators to have the margin? /cc @rickybanister

hypest commented:
There's an issue with the "Remove" button: If you rotate the device, the button no longer initiates the process.

While at it, the confirmation dialog is dismissed if the device is rotated.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/people/PeopleManagementActivity.java:
```diff
@@ -120,4 +124,113 @@ public void onPersonSelected(Person person) {
+                        String text;
+                        if (person != null) {
+                            PeopleTable.deletePerson(personID, localTableBlogId);
+                            text = getString(R.string.person_removed, person.getUsername());
```
Not sure if this is by design but, since we are asking for remove confirmation using the Person's display name, I think it makes sense to use the same in the Toast as well. Cheers!

oguzkocer replied:
It's by design. Although I agree with you, we want to match what Calypso does, and Calypso uses the exact same strings.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/people/PeopleManagementActivity.java:
```diff
@@ -120,4 +124,113 @@ public void onPersonSelected(Person person) {
+        });
+    }
+
+    private void confirmRemovePerson() {
```
It seems that we're not checking for current user when removing. Should that be allowed? If not, perhaps we can put out a message explaining why it cannot be done, or more simply remove the "remove" button completely. Cheers!

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/people/PeopleManagementActivity.java:
```diff
@@ -15,23 +18,31 @@
     private static final String KEY_PEOPLE_LIST_FRAGMENT = "people-list-fragment";
     private static final String KEY_PERSON_DETAIL_FRAGMENT = "person-detail-fragment";
 
+    private Person mSelectedPerson;
```
Would you consider keeping this state variable inside the PersonDetailFragment and fetch it from there when needed? Cheers!

This will also probably take care of the rotation problem (where the remove button doesn't do anything after rotating).

hypest commented:
A small difference is that Calypso uses the shortened "Admin" as the role _label_ underneath the user's name (both in user's list and in user details screen) instead of the full "Administrator" text we use here in the PR. Not important but perhaps we can address that in the spirit of following Calypso. Cheers!

hypest commented:
There is an "issue" with the hiding of the `People` row in MySiteFragment. Not in the changeset of this PR ofc but, it can lead to user confusion: The site info is updated in a rate-limited fashion, no sooner than 6hours (https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/WordPress.java#L130). That means that if a the app user is promoted to Admin, the updated capability (`list_users`) won't be available until 6hours have passed since the last sites update call. So, the delay can be up to 6hours, leaving the app user in confusion as to why she cannot see the `People`'s row in the app.

I think that this issue is not a blocker for this PR but is at least something we will need to keep in mind or even document for Support purposes.

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/people/PersonDetailFragment.java:
```diff
@@ -78,7 +136,7 @@ public void onResume() {
         if (!isAdded()) return;
 
-        Person person = PeopleTable.getPerson(mPersonID, mLocalTableBlogID);
+        Person person = getCurrentPerson();
```
Nitpicking: Let's lose the "Current" from the function name as it has some connotations (like, a globally selected "current" user) beyond the actual usage here. What do you think about "loadPerson()"? Cheers!

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/people/PersonDetailFragment.java:
```diff
@@ -175,17 +174,22 @@ public void onClick(View v) {
     }
 
+    @SuppressWarnings("deprecation")
+    // Remove the selectableItemBackground if the user can't be edited
```
May I suggest reinstating this comment back to the line the function is called? The function name is pretty clear and perfectly corresponding to its body, whereas the "if the user can't be edited" part is not implemented here but rather, at the caller's position. Sorry for the mind-blowing nitpicking! 😃 ❤️

hypest commented:
@oguzkocer , thanks for the effort on this one! I like the end result. I've finished my review pass at this point.

The only issue left, from my point of view, is the dismissal and re-opening of the role popup that doesn't always show the currently active role. Lemme know if I should provide more info on that and whether I should make another review pass. Cheers!

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/people/PersonDetailFragment.java:
```diff
@@ -106,4 +159,116 @@ public void setPersonDetails(long personID, int localTableBlogID) {
+    }
+
+    @SuppressWarnings("deprecation")
+    // Remove the selectableItemBackground if the user can't be edited
```
May I suggest reinstating this comment back to the line the function is called (line 177 at this point)? The function name is pretty clear and perfectly corresponding to its body, whereas the "if the user can't be edited" part is not implemented here but rather, at the caller's position. Sorry for the mind-blowing nitpicking! 😃 ❤️

oguzkocer commented:
> There is an "issue" with the hiding of the People row in MySiteFragment. Not in the changeset of this PR ofc but, it can lead to user confusion: The site info is updated in a rate-limited fashion, no sooner than 6hours (https://github.com/wordpress-mobile/WordPress-Android/blob/develop/WordPress/src/main/java/org/wordpress/android/WordPress.java#L130). That means that if a the app user is promoted to Admin, the updated capability (list_users) won't be available until 6hours have passed since the last sites update call. So, the delay can be up to 6hours, leaving the app user in confusion as to why she cannot see the People's row in the app.

> I think that this issue is not a blocker for this PR but is at least something we will need to keep in mind or even document for Support purposes.

I don't think this is a big deal in the sense that we have a lot of similar issues in the app. Ideally I agree that it shouldn't take that much time, but with the current architecture and this being a very edge case, most likely never even going to come up, I think it's fine to leave it as is. I'll still add this in the wrap-up post and give a heads up to Happiness Engineers. Thanks!

oguzkocer commented:
@hypest I think I've addressed all the comments & issues so far. If you like the fixes for the issues you've raised I think the only thing left is figuring out the design issue you've pinged @rickybanister about. Thanks!

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/people/RoleChangeDialogFragment.java:
```diff
@@ -0,0 +1,160 @@
+        Bundle args = getArguments();
+        if (args != null) {
+            String role = args.getString(ROLE_TAG);
+            mRoleListAdapter.setSelectedRole(role);
```
Let's also cover the case where the device is rotating with the dialog open. In the usual drill, we can use the `savedInstanceState` to pass the currently selected Role. Cheers!

hypest commented on WordPress/src/main/res/values/roles.xml:
```diff
@@ -1,8 +1,8 @@
+    <string-array name="roles" translatable="true">
         <item>
-            Administrator
+            Admin
```
This introduces the discrepancy where the role name does not follow the same copy as Calypso. That is, Calypso uses "Administrator" for the name and only the abbreviated "Admin" for the role "badge".

hypest commented:
@oguzkocer I finished my current pass. The only "issue" now is the rotation of the role dialog not keeping the selection and a quite minor thing with the "Administrator" text.

Besides those, the design issue you mentioned (same styling of the person name and role fields) is not a blocker from my point of view. We can raise it as an issue and continue the discussion there.

rickybanister commented:
An alternative design would be to use the header in a bigger way like this:

<img width="429" alt="screen shot 2016-05-16 at 11 57 14 am" src="https://cloud.githubusercontent.com/assets/437258/15296407/47d4927c-1b63-11e6-9c18-b68ab475a440.png">

hypest commented on WordPress/src/main/java/org/wordpress/android/ui/people/RoleChangeDialogFragment.java:
```diff
@@ -23,6 +23,15 @@
+    public void onSaveInstanceState(Bundle outState) {
+        super.onSaveInstanceState(outState);
+        String role = mRoleListAdapter.getSelectedRole();
+        if (role != null) {
```
Being extra careful doesn't hurt but there is actually no need to check for null at this point. For simplicity's sake, I'd recommend removing the check as there is no semantic value in it.

I would even go one step further and say that you can also remove the null-check around line 136 (`radioButton.setChecked(mSelectedRole.equalsIgnoreCase(role)`) by reversing the `equals` to `role.equalsIgnoreCase(mSelectedRole)` as the `role` is never expected to be null. Cheers!

oguzkocer replied:
You're right about the `line 136` it makes it much better to do it the way you've suggested. However for this line, `role` could actually be empty if it wasn't passed during initialization and the user hasn't selected anything yet. So, I think it's good to make sure this is not `null` before writing into the bundle.

hypest commented:
Thanks @rickybanister , that alternative design looks interesting and I can see it solving the styling issue I brought up! I'd most certainly want to see it in action as well!

As mentioned earlier, I will not block this PR over this particular issue but I'll leave it to @oguzkocer to decide whether to give it a stab as part of this PR or defer it to a subsequent one. Cheers!

rickybanister commented:
I think it would be fine to merge this pr and work on that design in a new one.
